File: Jan2021/04Jan/DreamoonAndWiFi.java
/*
Problem Name: Dreamoon and WiFi
Problem Link: http://codeforces.com/problemset/problem/476/B
*/
import java.util.*;
public class MyClass {
    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        String s1 = in.next(), s2 = in.next();
        int original_pos = 0;
        for(char c:s1.toCharArray())
            original_pos += (c == '+' ? 1 : -1);
        int ques = 0, supposed_pos = 0;
        for(char c:s2.toCharArray()){
            ques += (c == '?' ? 1 : 0);
            supposed_pos += (c == '+' ? 1 : -1);
        }
        if(ques == 0){
            if(supposed_pos == original_pos)
                System.out.println(1d);
            else
                System.out.println(0d);
        }
        else{
            double prob = 0;
            int possibilities = (int)Math.pow(2,ques), temp, position;
            for(int i=0;i<possibilities;i++){
                temp = i;
                position = 0;
                for(char c:s2.toCharArray()){
                    if(c == '?'){
                        if((temp & 1) == 0)
                            position++;
                        else
                            position--;
                        temp >>= 1;
                    }
                    else{
                        position += (c == '+' ? 1 : -1);
                    }
                }
                if(position == original_pos)
                    prob++;
            }
            System.out.println(prob/possibilities);
        }
    }
}

File: Jan2021/04Jan/DreamoonAndWiFi.cpp
/*
Problem Name: B. Dreamoon and WiFi
Problem Link: http://codeforces.com/problemset/problem/476/B
*/

#include<bits/stdc++.h>
using namespace std;

#define test(t)     int t; cin>>t; while(t--)
#define ll          long long

int main(){
    ios::sync_with_stdio(0);    
    cin.tie(0); 
    string s1,s2;
    ll actual=0, uk=0, current;
    double ans;
    
    getline(cin, s1);
    getline(cin, s2);
    
    for(int i=0;i<s1.length();++i){
        if(s2[i]=='?')
            uk++;
        if(s1[i]=='+')
            actual++;
        else
            actual--;
        if(s2[i]=='+')
            actual--;
        else if(s2[i]=='-')
            actual++;
    }
    
    if(uk==0){
        ans = (actual == 0)? 1: 0;
    }
    else{
        ll total=pow(2, uk), rem, curr, prob=0;
        for(ll i=0;i<total;++i){
            rem=actual;
            curr=i;
            for(ll j=0;j<uk;++j){
                if((curr & 1) == 0)
                    rem--;
                else
                    rem++;
                curr>>=1;
            }
            if(rem==0){
                prob++;
            }
        }
        ans=(double)prob/total;
    }
    cout<<fixed<<setprecision(12)<<ans;
    
    return 0;
}
